synapse/storage/database.py

```diff
@@ -1281,20 +1291,33 @@ def simple_upsert_many_txn_native_upsert(
                 k + "=EXCLUDED." + k for k in value_names
             )
 
-        sql = "INSERT INTO %s (%s) VALUES (%s) ON CONFLICT (%s) DO %s" % (
-            table,
-            ", ".join(k for k in allnames),
-            ", ".join("?" for _ in allnames),
-            ", ".join(key_names),
-            latter,
-        )
-
         args = []
 
         for x, y in zip(key_values, value_values):
             args.append(tuple(x) + tuple(y))
 
-        return txn.execute_batch(sql, args)
+        if isinstance(txn.database_engine, PostgresEngine):
+            # We use `execute_values` for postgres as it can be a lot faster
+            # than `execute_batch`, but it's only available on postgres.
+            sql = "INSERT INTO %s (%s) VALUES ? ON CONFLICT (%s) DO %s" % (
+                table,
+                ", ".join(k for k in allnames),
+                ", ".join(key_names),
+                latter,
+            )
+
+            txn.execute_values(sql, args, fetch=False)
+
+        else:
```

Comment:
I thought sqlite doesn't have native upsert? So will this ever get called?

Reply:
It can on sqlite 3.24+ :D

https://github.com/matrix-org/synapse/blob/00640ee71affeab0f6ed7e784925522437f076c5/synapse/storage/engines/sqlite.py#L50-L56

Reply:
Oh, didn't realize that! 🎉